### PR TITLE
Updated Core Protocol in preparation of accounts support (Closes #603)

### LINKED ISSUE #603: Create base contract protocol
### Problem

The way our contract chain messages are designed right now needs to be modified because they do not handle protocol upgrades or key management (#92). Of these two issues, handling protocol upgrades is the most urgent.

An example of a protocol upgrade is a new voting rule type being added to the GroupContract that didn't exist before.

**EDIT:** Additionally, we currently have no way to distinguish between publicly-visible and privately-visible data for a contract. For example, it would be useful for someone who wants to join a group to see what the group's profile picture is (if they set its visibility public), or other information about the group. This should probably be split out into a separate issue.

### Solution

```
    //       define a base set of protocol messages that are publicly readable
    //       OP_CONTRACT - create a contract with a given name (publicly readable),
    //                     and authorized keys, but all other data is encrypted.
    //       OP_ACTION - an action is applied to the contract. its name and its
    //                   data is encrypted.
    //       OP_KEY_* - key related ops that determine who can write to this contract
    //       OP_PROTOCOL_UPGRADE - bump the protocol version, clients that are less
    //                             than this version cannot read or write
```

At runtime, clients in a group will support the lowest protocol version supported by any group member only. This will be determined by looking at the lowest protocol version of all member IdentityContracts, and using that as the supported protocol version for the GroupContract. As clients upgrade, new features can become available. This ensures that all members are able to process all messages. New members cannot join a group unless their client supports its protocol version (which can be updated by anyone in the group - but see important note in #602).

Useful links:

- [dynamic `import()`](https://rollupjs.org/guide/en/#importing)
- https://vuex.vuejs.org/api/#hotupdate
- https://vuex.vuejs.org/guide/modules.html#preserving-state

**EDIT:** regarding the publicly-visible data issue in the above edit, we could use `OP_PROP_SET` to selectively reveal data by setting public properties on the contract, and reading those.

**EDIT:** per [@sandrina-p's comment](https://github.com/okTurtles/group-income-simple/pull/611#discussion_r313870216), modify `GIMessage.create` to take an object as parameters instead of a long list of unnamed parameters.

### PR DESCRIPTION
**Pull request by taoeffect**
### Upgraded Core Protocol (Closes #603)

This PR includes a major change to the design of the core Group Income protocol in preparation for full accounts support and end-to-end encryption.

The core contract class `GIMessage` has been modified to add the concept of operations. The following operations are supported:

- `OP_CONTRACT` — Creates a new contract chain
- `OP_ACTION` — Performs an end-to-end encrypted action on a contract (e2e cryptography not yet implemented)
- `OP_PUB_ACTION` — Performs a publicly-readable action on a contract
- `OP_KEY_AUTH` — Authorizes a new key to be able to write to a contract
- `OP_KEY_DEAUTH` — Invalidates a key (not implemented)
- `OP_PROTOCOL_UPGRADE` — Advertises a new version of the protocol (not implemented)
- `OP_PROP_SET` — Sets a publicly readable key-value pair on the contract
- `OP_PROP_DEL` — Removes a publicly readable key-value pair from the contract

One of the significant changes that has resulted from this is that the very first message to a contract is no longer the contract's constructor action. Instead, the constructor action is posted second in the chain. The first message merely advertises the existence of a new contract via `OP_CONTRACT`, and the constructor comes next via `OP_ACTION` or `OP_PUB_ACTION`.

### Miscellaneous Changes

- We're now using native browser modules instead of SystemJS. `systemjs` dependency removed.
- Removed `babel-plugin-transform-inline-environment-variables` and replaced it with `@rollup/plugin-replace`
- Re-enabled the babel cache
- Build process now has two entry points and two outputs: one for the main bundle, and another for a service worker (which is currently unused in the app). As part of this, our build process now correctly supports multiple inputs/outputs via rollup.
- Ran `npm audit fix` to update some deep dependencies
- Removed `fromPairs` in `giLodash` (use [`Object.fromEntries`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries)) and added `mapValues`

### CODE REVIEW

File: backend/server.js
Comment (snowteamer) [inline backend/server.js:32]: I'd rather start with a minimal set of origins, then whitelist additional origins later as necessary. However that method would require us to keep in mind that not all origins are accepted.

File: backend/server.js
Comment (snowteamer) [inline backend/server.js]: This expression could evaluate to `false` if it was no longer commented out, likely causing an error.

This can be avoided by using a ternary expression like:
```js
origin: (
  process.env.NODE_ENV === 'development' ?
  [process.env.API_URL, 'http://localhost:3000'] :
  [process.env.API_URL]
)
```
or the drier, but less readable:
```js
origin: [
  process.env.API_URL,
  ...(process.env.NODE_ENV === 'development' ? ['http://localhost:3000'] : []),
]
```

File: frontend/controller/backend.js
Comment (snowteamer) [inline frontend/controller/backend.js]: Isn't `attempt` always equal to `maxAttempts` here, and therefore redundant?

File: frontend/model/contracts/group.js
Comment (snowteamer) [inline frontend/model/contracts/group.js:240]: If this group has no defined mincome currency, not even a default one like the USD, then calling this function is probably an error which should be reported.

File: frontend/model/contracts/group.js
Comment (snowteamer) [inline frontend/model/contracts/group.js:514]: Nit: I would prefer the name `contractsByID` here, since it does not refer to an array but a plain object.

File: frontend/model/state.js
Comment (snowteamer) [inline frontend/model/state.js]: This function returns a number but its name looks very much like a boolean function name.

I would rather suggest `countPartial()`, or `countPartialPayments()`, or `getNumberOfPartialPayments()`, since it finds the number of partial payments in a list.

File: frontend/model/state.js
Comment (snowteamer) [inline frontend/model/state.js:357]: I'm not sure about returning undefined here, it's probably unsafe. It also makes the function"s return type permanently more complex to handle an edge case which should be exceptional. So I would favor throwing instead.

File: frontend/model/state.js
Comment (snowteamer) [inline frontend/model/state.js:526]: Same point here

File: frontend/model/state.js
Comment (snowteamer) [inline frontend/model/state.js:579]: Another suspicious return statement. If there are no members in the group, it's ambiguous whether their profiles are sorted, so it would be safer to avoid the question altogether if possible.

File: shared/GIMessage.js
Comment (snowteamer) [inline shared/GIMessage.js:40]: - Abbreviations in in public APIs should generally be avoided.
  For example, the PUB part in `OP_PUB_ACTION` is very easy to read as PUBLISH instead of PUBLIC.
- When sticking to abbreviations, they should arguably be used consistently.
  In this case, PROTOCOL could be abbreviated as PROTO.
- Since `OP_ACTION` means an encrypted action according to the comments,
  Shouldn't it be written `OP_ENCRYPTED_ACTION` to be more explicit,
  or PRIVATE_ACTION in contrast with `OP_PUB_ACTION`?
- Some operation identifiers are nouns (e.g. action, contract) while other are verbs (prop_set). Some can also be read as both (auth, update).
  Would it be possible to use only names or only verbs to improve consistency?

File: shared/GIMessage.js
Comment (pieer) [inline shared/GIMessage.js:130]: Could initialize directly without push here

File: shared/GIMessage.js
Comment (pieer) [inline shared/GIMessage.js:153]: Same

File: shared/GIMessage.js
Comment (pieer) [inline shared/GIMessage.js]: forn ow?

**Review by pieer — DISMISSED** [review]
Great work @taoeffect. I could only find very tiny detail to maybe change.
Approved!

File: frontend/controller/backend.js
Comment (snowteamer) [inline frontend/controller/backend.js]: Isn't `attempt` always equal to `maxAttempts` here? If so, no need to print it twice in the error message.

File: frontend/index.html
Comment (snowteamer) [inline frontend/index.html]: Is using a dynamic import statement better than using `type=module` together with a `src` attribute?

File: frontend/model/contracts/Contract.js
Comment (snowteamer) [inline frontend/model/contracts/Contract.js:11]: So "Chelonia" is the name of the GI protocol?

Well that's cute. I like it! 😄

File: frontend/model/contracts/group.js
Comment (snowteamer) [inline frontend/model/contracts/group.js:240]: This will return a falsy value if the group has no mincome currency.

And in that case, calling this function doesn't make sense anyway, therefore throwing an error seems appropriate.

File: frontend/model/contracts/group.js
Comment (snowteamer) [inline frontend/model/contracts/group.js:514]: I would find it easier to remember that `contracts` is not an array but actually a plain object whose keys are contract IDs, if it was named `contractsByID` instead.

File: frontend/model/state.js
Comment (snowteamer) [inline frontend/model/state.js:357]: I would favor either returning an empty object or throwing an error here.

File: frontend/model/state.js
Comment (snowteamer) [inline frontend/model/state.js:526]: Same point here also.

File: frontend/model/state.js
Comment (snowteamer) [inline frontend/model/state.js:579]: If there are currently no members in the group, it's ambiguous whether the member list is sorted.

So it could be safer to avoid the question altogether, by avoiding to call the function in that case in the first place, and by having it throw an exception if it does get called.

**Review by snowteamer — DISMISSED** [review]
Congratulations for the hard work on this one!

I have still a few files to review, namely the tests. So I will likely add a few more comments later, but am already approving so as not to block the merge process.

File: shared/GIMessage.js
Comment (taoeffect) [reply shared/GIMessage.js:130]: Do you mean something like this?

```js
if (!state._vm.authorizedKeys) {
  state._vm.authorizedKeys = [{ key: v.authkey, context: 'owner' }]
  return
}
```

That is a few extra lines and it would violate DRY since `{ key: v.authkey, context: 'owner' }` is repeated, so I chose to do it this way

File: frontend/model/state.js
Comment (taoeffect) [reply frontend/model/state.js:579]: So, these getters are attached to UI elements... and with the modifications I made in this PR they will throw an exception and that breaks the rendering of the Vue components that bind to the getters. This return statement prevents the exception from occurring and allows the UI to render.

Cleaner versions of this that accomplish the same thing are certainly welcome!

File: shared/GIMessage.js
Comment (taoeffect) [reply shared/GIMessage.js]: Oops. Meant "nothing for now", nice catch, will be fixed in the next commit, thanks!

File: frontend/model/state.js
Comment (taoeffect) [reply frontend/model/state.js:357]: The UI was breaking because an exception was thrown, so I returned `undefined` here to fix it. The issue with returning an empty object is that it's not clear to me personally whether or not `{}` is a falsey value. Even if it is, that requires some deep JS knowledge to know. And other getters might rely on the return value of this to perform additional computations. This way they can also check if this getter returned a false value and return immediately without throwing an exception that breaks UI rendering.

File: frontend/model/contracts/group.js
Comment (taoeffect) [reply frontend/model/contracts/group.js:514]: Hmm, this would be a complex refactor because the project has so many instances of "contracts", so if you'd like this change to be made, can you open an issue for it?

File: frontend/model/contracts/group.js
Comment (taoeffect) [reply frontend/model/contracts/group.js:240]: Perhaps in this case that does make sense for this getter... I'll add a TODO here for now.

File: frontend/model/contracts/Contract.js
Comment (taoeffect) [reply frontend/model/contracts/Contract.js:11]: Glad you like it! It's a working name, and might change... will probably bring it up on Slack to hear people's opinions about this name and another I'm considering.

File: frontend/index.html
Comment (taoeffect) [reply frontend/index.html]: Interesting, I've no clue which is better, but your suggestion looks better so I'll use that in a followup commit!

File: backend/server.js
Comment (taoeffect) [reply backend/server.js:32]: Well, eventually we want servers to talk to each other ... so... 🤷‍♂️

File: shared/GIMessage.js
Comment (taoeffect) [reply shared/GIMessage.js:40]: Think this should be resolved now? I left the PROPs as the same because I couldn't think of anything better to replace them with.

**Review by CrazedCoding — APPROVED** [review]
It all looks well written, reasonably efficient, and logically air-tight, ALTHOUGH I think I need to run some tests: I am very interested in seeing how, if at all, updates to the core protocol are going to affect the mocha tests I am making a PR for.

File: backend/server.js
Comment (snowteamer) [reply backend/server.js:32]: In that case, I don't know if it's even possible to not use `'*'`

File: frontend/model/state.js
Comment (snowteamer) [reply frontend/model/state.js:357]: Alright, let's leave it as is for now. There might be something to improve in the calling UI code and getters, but it can be fixed in a separate issue.

File: frontend/model/contracts/group.js
Comment (snowteamer) [reply frontend/model/contracts/group.js:514]: If you think it's a good idea, then I'd be glad to (when I have time)

File: frontend/model/state.js
Comment (snowteamer) [reply frontend/model/state.js:579]: Thank you for the clarifications!

File: shared/GIMessage.js
Comment (snowteamer) [reply shared/GIMessage.js:40]: I would tend to misread PROP as property, or proposal, while it actually means a public key/value pair according to the inline comments.
So maybe `OP_KEYPAIR_SET` and `OP_KEYPAIR_DELETE`? Though, these names still leave the "public" part out - but iI might be descriptive enough if there are no operations involving private keypairs.

File: shared/GIMessage.js
Comment (snowteamer) [inline shared/GIMessage.js:84]: The name `decrypted` sounds like a boolean or boolean function to me, but it returns a decrypted op value (correct me if I am wrong); so what about `decryptedValue` ?

**Review by snowteamer — APPROVED** [review]

File: test/backend.test.js
Comment (snowteamer) [inline test/backend.test.js:106]: This statement is confusing, why is the return value not used? Is it just to check that the call doesn't throw?

File: shared/GIMessage.js
Comment (taoeffect) [reply shared/GIMessage.js:84]: OK, can make that change in a followup PR so as to not force another round of review!

File: test/backend.test.js
Comment (taoeffect) [reply test/backend.test.js:106]: In order to perform the decryption on the message and store the value in `this._decrypted` for followup invocations via `msg.decrypted()` (or in your case, `msg.decryptedValue()`).

**Pull request merged**
